jenkins-dsl/jenkins_jira_integration.py

```diff
@@ -0,0 +1,352 @@
+
+        self.url = build['url']
+
+        previous_build = build['previousBuild']
+        if previous_build is None:
+            self.previous_number = NO_PREVIOUS_BUILD
+        else:
+            self.previous_number = previous_build[NUMBER]
+
+        self.result = build['result']
+
+        tests = server.get_build_test_report('Cassandra-' + branch, int(build_num), depth=5)
+
+        self.passcount = 0
+        self.failcount = 0
+        self.test_failures: Set[str] = set()
+
+        # If we had a bad build, it's possible we have no test run results and should just be done with it
+        if not tests:
+            return
+
+        self.passcount = tests['passCount']
+        self.failcount = tests['failCount']
+
+        suites = tests['suites']
+        for suite in suites:
+            for case in suite['cases']:
+                # We treat failures and regressions the same as we're going to rely on our history cache to provide per-test failure context
+                if (case['status'] == 'FAILED') or case['status'] == 'REGRESSION':
+                    self.test_failures.add(case['className'] + '.' + case['name'])
+
+    def string_detailed(self) -> str:
+        """
+        Builds out detailed job runs for posting on JIRA. Whitespace is lost but it looks clean enough w/things left
+        aligned excepting the affected paths as a bullet list.
+        """
+        result = ''
+        result += 'Branch: ' + self.branch + ', build number: ' + str(self.number) + '\n'
+        result += '   jenkins url: ' + str(self.url) + '\n'
+        result += '   JIRA: CASSANDRA-' + str(self.JIRA) + '\n'
+        result += '   commit url: https://git-wip-us.apache.org/repos/asf?p=cassandra.git;a=commit;h=' + str(self.sha) + '\n'
+        result += '   affected paths:' + '\n'
+        for path in self.paths:
+            result += '* ' + path + '\n'
+        result += '\n   Build Result: ' + self.result + '\n'
+        result += '   Passing Tests: ' + str(self.passcount) + '\n'
+        result += '   Failing Tests: ' + str(self.failcount) + '\n'
+        return result
+
+
+def retrieve_build_details(server: jenkins, branch: str, build_num: int) -> BuildData:
+    result = None
+    found_build = False
+    while not found_build:
+        log('Retrieving requested build: ' + str(build_num) + '. Branch: ' + branch + '. build_num: ' + str(build_num))
+        result = BuildData(server, branch, str(build_num))
+
+        # Some commit messages aren't JIRA related (ninjas, deb release changes, etc)
+        if result.JIRA == UNKNOWN:
+            print('No related CASSANDRA-NNNNN Jira found for build: ' + str(result.number) + '. Commit Message: ' + result.commit_msg)
+            print('Exiting processing; nothing to be done for this build if we can\'t determine the JIRA ticket it\'s associated with.')
+            sys.exit(-1)
+
+        log('Parsed JIRA number: ' + result.JIRA + ' for build: ' + str(result.number))
+        break
```

Comment:
is this not `while True`?  I don't see you updating `found_build`

Reply:
Just removed loop entirely; used to walk until I found next valid build to compare against as initial model was 1:1 comparison of build vs. last. Now do history, so this method is really about just parsing this one build and seeing if we can infer Jira from it.